### PR TITLE
Added week 3 code samples

### PR DESCRIPTION
**Pull request by haydenhur**

### CODE REVIEW

File: 03/1_pointer/main.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    int max_val;
+    
+    get_min_max(nums, LENGTH, &min_val, &max_val);
+    
+    printf("min_val: %d\n", min_val);
+    printf("max_val: %d\n", max_val);
+    
+    return 0;
+}
+
```
Comment (popekim) [inline 03/1_pointer/main.c:21]: add precondition saying length is at least 1.

then add assert in the function to check length is at least 1

File: 03/1_pointer/main.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#include <stdio.h>
+
+void get_min_max(const int nums[], const size_t length, int* out_min, int* out_max);
+
+int main(void)
+{
+    int nums[5] = {3, 10, 9, 8, 7};
```
Comment (popekim) [inline 03/1_pointer/main.c]: { 3, 10, 9, 8, 7 }

Comment (popekim) [reply 03/1_pointer/main.c]: const int

File: 03/1_pointer/main.c [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+int main(void)
+{
+    int nums[5] = {3, 10, 9, 8, 7};
+    const size_t LENGTH = 5u;
+    int min_val;
+    int max_val;
+    
+    get_min_max(nums, LENGTH, &min_val, &max_val);
+    
+    printf("min_val: %d\n", min_val);
```
Comment (popekim) [inline 03/1_pointer/main.c]: "min:
"nax:

File: 03/2_memory_viewer/main.c [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#include <stdio.h>
+
+int main(void)
+{
+    int num = 0x12345678;
+    char* num_address = (char*)&num;
+    size_t i = 0;
```
Comment (popekim) [inline 03/2_memory_viewer/main.c]: size_t i;

File: 03/2_memory_viewer/main.c [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#include <stdio.h>
+
+int main(void)
+{
+    int num = 0x12345678;
+    char* num_address = (char*)&num;
```
Comment (popekim) [inline 03/2_memory_viewer/main.c]: const char* p =

File: 03/2_memory_viewer/main.c [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    int num = 0x12345678;
+    char* num_address = (char*)&num;
+    size_t i = 0;
+    
+    for(i = 0; i < sizeof(num); ++i) {
+        printf("%hhx ", *(num_address + i));
+    }
+    printf("\n");
+    
+    printf("NUM in hex form: %x", NUM);
```
Comment (popekim) [inline 03/2_memory_viewer/main.c]: i don't see NUM.

does this even compile?

File: 03/3_is_array_overlap/main.c [hunk 7]
```diff
@@ -0,0 +1,34 @@
+#include <stdio.h>
+
+int is_overlap(int nums1[], const size_t length1, int nums2[], const size_t length2);
+
+int main(void)
+{
+    int nums1[5] = {1, 2, 3, 4, 5};
```
Comment (popekim) [inline 03/3_is_array_overlap/main.c]: { 1...

File: 03/3_is_array_overlap/main.c [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#include <stdio.h>
+
+int is_overlap(int nums1[], const size_t length1, int nums2[], const size_t length2);
+
+int main(void)
+{
+    int nums1[5] = {1, 2, 3, 4, 5};
+    int nums2[5] = {1, 2, 8, 9, 10};
```
Comment (popekim) [inline 03/3_is_array_overlap/main.c]: { 1 ...

File: 03/3_is_array_overlap/main.c [hunk 9]
```diff
@@ -0,0 +1,34 @@
+#include <stdio.h>
+
+int is_overlap(int nums1[], const size_t length1, int nums2[], const size_t length2);
+
+int main(void)
+{
+    int nums1[5] = {1, 2, 3, 4, 5};
+    int nums2[5] = {1, 2, 8, 9, 10};
+    int* nums3 = nums1;
```
Comment (popekim) [inline 03/3_is_array_overlap/main.c]: bad example. 

better example will be

int nums1[10];
int nums2[5];

int* num3 = nums1 + 2; // and length is like 5

File: 03/3_is_array_overlap/main.c [hunk 10]
```diff
@@ -0,0 +1,34 @@
+#include <stdio.h>
+
+int is_overlap(int nums1[], const size_t length1, int nums2[], const size_t length2);
+
+int main(void)
+{
+    int nums1[5] = {1, 2, 3, 4, 5};
+    int nums2[5] = {1, 2, 8, 9, 10};
+    int* nums3 = nums1;
+    const size_t LENGTH = 5u;
```
Comment (popekim) [inline 03/3_is_array_overlap/main.c]: if you are going to make this as const.. u should use this for array initialization as well..

if not sizeof(nums1) / sizeof(nums1[0]) is a better way 

or make a macro for this #define ARRAY_LENGTH(arr) (sizeof(arr) / sizeof(arr[0])

File: 03/3_is_array_overlap/main.c [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+int is_overlap(int nums1[], const size_t length1, int nums2[], const size_t length2)
+{
+    size_t i;
+ 
+    if (length1 != length2) {
+        return 0;
+    }
+    for (i = 0; i < length1; ++i) {
```
Comment (popekim) [inline 03/3_is_array_overlap/main.c]: do it without for loop.

nums1's boundary: [nums1, nums1 + length)
nums2's boundary: [nums2, nums2 + length)

File: 03/3_is_array_overlap/main.c [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        (is_overlap(nums1, LENGTH, nums3, LENGTH) ? "Yes" : "No"));
+    
+    return 0;
+}
+
+int is_overlap(int nums1[], const size_t length1, int nums2[], const size_t length2)
+{
+    size_t i;
+ 
+    if (length1 != length2) {
```
Comment (popekim) [inline 03/3_is_array_overlap/main.c]: simply wrong

File: 03/4_const_pointer/main.c [hunk 13]
```diff
@@ -0,0 +1,23 @@
+#include <stdio.h>
+
+void add_vec3(const int* v1, const int* v2, int* out_v3);
+
+int main(void)
+{
+    int v1[3] = {1, 2, 3};
```
Comment (popekim) [inline 03/4_const_pointer/main.c]: const int v1,
const int v2
int out_v3

File: 03/1_pointer/main.c [hunk 14]
```diff
@@ -0,0 +1,35 @@
+#include <stdio.h>
+
+void get_min_max(const int nums[], const size_t length, int* out_min, int* out_max);
```
Comment (popekim) [inline 03/1_pointer/main.c]: why are we not making separate h and c files for all the sample functions now?

File: 03/5_pointer_array/main.c [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    int* nums2[3];
+    
+    nums2[0] = nums[0];
+    nums2[1] = nums[1];
+    nums2[2] = nums[2];
+    
+    printf("nums[0] address: %p\n", (void*)nums[0]);
+    printf("nums[1] address: %p\n", (void*)nums[1]);
+    printf("nums[2] address: %p\n", (void*)nums[2]);
+    printf("offset nums[2] to nums[0]: %d\n", nums[2] - nums[0]);
```
Comment (popekim) [inline 03/5_pointer_array/main.c]: nums[2]'s offset from nums[0]

File: 03/5_pointer_array/main.c [hunk 16]
```diff
@@ -0,0 +1,31 @@
+#include <stdio.h>
+
+int main(void)
+{
+    int nums[3][5] = {
```
Comment (popekim) [inline 03/5_pointer_array/main.c]: const?

File: 03/5_pointer_array/main.c [hunk 17]
```diff
@@ -0,0 +1,31 @@
+#include <stdio.h>
+
+int main(void)
+{
+    int nums[3][5] = {
+        { 35, 50, 65, 24, 71 }, 
+        { 52, 93, 30, 18, 83 },
+        { 72, 53, 41, 88, 69 }
+    };
+    int* nums2[3];
```
Comment (popekim) [inline 03/5_pointer_array/main.c]: const int* pointer_array

File: 03/5_pointer_array/main.c (hunk 15 above)
Comment (popekim) [reply 03/5_pointer_array/main.c]: show what does it output as comments

**Review by popekim — CHANGES_REQUESTED** [review]

File: 03/1_pointer/main.c [hunk 18]
```diff
@@ -0,0 +1,35 @@
+#include <stdio.h>
+
+void get_min_max(const int nums[], const size_t length, int* out_min, int* out_max);
+
+int main(void)
+{
+    int nums[5] = {3, 10, 9, 8, 7};
+    const size_t LENGTH = 5u;
+    int min_val;
```
Comment (monkeywithbanana) [inline 03/1_pointer/main.c]: min/max no need for `_val`

File: 03/3_is_array_overlap/main.c [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+    int nums2[5] = {1, 2, 8, 9, 10};
+    int* nums3 = nums1;
+    const size_t LENGTH = 5u;
+    
+    printf("Are nums1 and nums2 overlapped?: %s\n", 
+        (is_overlap(nums1, LENGTH, nums2, LENGTH) ? "Yes" : "No"));
+    printf("Are nums1 and nums3 overlapped?: %s\n", 
+        (is_overlap(nums1, LENGTH, nums3, LENGTH) ? "Yes" : "No"));
+    
+    return 0;
```
Comment (monkeywithbanana) [inline 03/3_is_array_overlap/main.c:21]: for true and false, define macros TRUE (1) FALSE (0)

**Review by monkeywithbanana — CHANGES_REQUESTED** [review]

File: 03/1_pointer/main.c [hunk 20]
```diff
@@ -0,0 +1,39 @@
+#include <assert.h>
+#include <stdio.h>
+
+#include "algorithm.h"
+
+int main(void)
```
Comment (popekim) [inline 03/1_pointer/main.c:8]: this time let's use

#define LENGTH (5)

and use it for both array init and size_t

File: 03/3_is_array_overlap/main.c [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+int main(void)
+{
+    int nums1[10] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+    int nums2[5] = { 1, 2, 8, 9, 10 };
+    int* nums3 = nums1 + 2;
+    const size_t NUMS3_LENGTH = 5u;
+
+    printf("Are nums1 and nums2 overlapped?: %s\n",
+        is_overlap(nums1, ARRAY_LENGTH(nums1), nums2, ARRAY_LENGTH(nums2)) == TRUE
```
Comment (popekim) [inline 03/3_is_array_overlap/main.c]: this code is getting too long. so i'd say move 

is_overlap part to a separate line.. assign the return value to char* .. and print that..

also remove (==TRUE) adding this on ternary op sounds weird

File: 03/4_const_pointer/main.c [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    add_vec3(v1, v2, v3);
+    
+    printf("v3: %d, %d, %d", v3[0], v3[1], v3[2]);
+
+    return 0;
+}
+
+void add_vec3(const int* v1, const int* v2, int* out_v3)
+{
+    *out_v3 = *v1 + *v2;
```
Comment (popekim) [inline 03/4_const_pointer/main.c]: for i to 3
*out_v3 = *v1 + *v2 이런식으로 증가
++out_v3
++v1
++v2

**Review by popekim — APPROVED** [review]

**Review by monkeywithbanana — APPROVED** [review]

**Pull request merged**